src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -64,17 +63,11 @@ internal partial class DiagnosticIncrementalAnalyzer : IIncrementalAnalyzer2
 
         internal DiagnosticAnalyzerInfoCache DiagnosticAnalyzerInfoCache => _diagnosticAnalyzerRunner.AnalyzerInfoCache;
 
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/54400", Constraint = "Avoid calling GetAllHostStateSets on this hot path.")]
         public bool ContainsDiagnostics(ProjectId projectId)
         {
-            foreach (var stateSet in _stateManager.GetStateSets(projectId))
-            {
-                if (stateSet.ContainsAnyDocumentOrProjectDiagnostics(projectId))
-                {
-                    return true;
-                }
-            }
-
-            return false;
+            return _stateManager.HasAnyHostStateSet(static (stateSet, arg) => stateSet.ContainsAnyDocumentOrProjectDiagnostics(arg), projectId)
+                || _stateManager.HasAnyProjectStateSet(projectId, static (stateSet, arg) => stateSet.ContainsAnyDocumentOrProjectDiagnostics(arg), projectId);
```

Comment:
i don't quite get why this pattern is preferred.  nowe we have some helpers, but they're only called from exactly this location.  and we need to pass static lambdas to it.  why not just inline both of those helpers here and just doc what is/isn't allowed?

Reply:
A bunch of types and members used in the helpers are not accessible from this location. I added some comments to the helpers, but kept them since it seemed cleaner than a rewrite that exposed the data.